fix(ivy): include type parameter for `ngBaseDef` declaration

JoostK opened the pull request:
When a class uses Angular decorators such as `@Input`, `@Output` and
friends without an Angular class decorator, they are compiled into a
static `ngBaseDef` field on the class, with the TypeScript declaration
of the class being altered to declare the `ngBaseDef` field to be of type
`ɵɵBaseDef`. This type however requires a generic type parameter that
corresponds with the type of the class, however the compiler did not
provide this type parameter. As a result, compiling a program where such
invalid `ngBaseDef` declarations are present will result in compilation
errors.

This commit fixes the problem by providing the generic type parameter.

Fixes #31160

JoostK committed:
fix(ivy): include type parameter for `ngBaseDef` declaration

When a class uses Angular decorators such as `@Input`, `@Output` and
friends without an Angular class decorator, they are compiled into a
static `ngBaseDef` field on the class, with the TypeScript declaration
of the class being altered to declare the `ngBaseDef` field to be of type
`ɵɵBaseDef`. This type however requires a generic type parameter that
corresponds with the type of the class, however the compiler did not
provide this type parameter. As a result, compiling a program where such
invalid `ngBaseDef` declarations are present will result in compilation
errors.

This commit fixes the problem by providing the generic type parameter.

Fixes #31160

petebacondarwin commented on packages/compiler/src/compiler_facade_interface.ts:
```diff
@@ -151,6 +151,7 @@ export interface R3ComponentMetadataFacade extends R3DirectiveMetadataFacade {
 
 export interface R3BaseMetadataFacade {
   name: string;
+  type: any;
```
What would break if this change was not here?

JoostK replied:
[This test](https://github.com/angular/angular/blob/6085f335e8137ad560223687c2cba80575569905/packages/compiler/test/compiler_facade_interface_spec.ts) verifies that the understanding of Angular metadata is compatible from core to compiler and vice versa.

JoostK replied:
Also, this would be a type error:

https://github.com/angular/angular/blob/f74373f2dd829249e77537a90883eda833a8ce85/packages/compiler/src/jit_compiler_facade.ts#L160-L168

petebacondarwin replied:
Cool. That's what I was looking for.

JoostK replied:
Haha, there appears to be a catch! There's no `R3BaseMetadataFacade` in the test file at all ;) I'll add it.

JoostK committed:
test(ivy): verify equality of ngBaseDef metadata between compiler and core

There's no build time dependency from @angular/core to @angular/compiler,
so core can't directly refer to compiler types. To overcome this limitation,
there's a facade layer defined in the compiler and duplicated in core,
such that during runtime all types will correctly align.

There's a testcase in the compiler that verifies that all such facade types
are compatible across core and compiler, such that the core types can't get
misaligned with the actual definitions in the compiler. This suite of
tests were missing the `R3BaseMetadataFacade` facade type, so it was possible
for this type to get out of sync.

JoostK commented:
merge-assistance: needs presubmit :-)

AndrewKushnir commented:
- [VE Presubmit](http://test/OCL:255987228:BASE:255987246:1562003282399:78d9d586)
- [Ivy Presubmit](https://test.corp.google.com/ui#id=OCL:255988373:BASE:255988523:1562003671949:e86210da)

AndrewKushnir commented:
VE Presubmit is successful, Ivy Presubmit looks normal. Thank you.
